### PR TITLE
Document HTTP-JSON APIs [PLEN-57]

### PR DESCRIPTION
**Pull request by miklos-da**
Adds documentation to HTTP-JSON APIs:
* Document common gRPC & HTTP metrics on a `Common Metrics` page.
* Reference common HTTP metrics from `trigger-service`'s and `json-api-service`'s page.
* Describe how to enable metrics reporting on `trigger-service`'s page.

### CODE REVIEW

File: docs/source/tools/trigger-service/index.rst [hunk 1]
```diff
@@ -352,3 +352,26 @@ HTTP Response
...
+
+Metrics
+*******
+
+Enable and Configure Reporting
+==============================
+
+To enable metrics and configure reporting, you can use the below config block in application config:
+
+.. code-block:: none
```
Comment (hrischuk-da) [inline docs/source/tools/trigger-service/index.rst:365]: Can we change this example to be 'Prometheus' so the user has an example for the format that we want them to use?

**Review by katmurp — COMMENTED** [review]
From a language perspective, it would be good to clarify the different between the two headers **Common HTTP Metrics** and **List of Metrics** if there is some significant difference between them. Or join them if there isn't. You can do this in the title, i.e. **Common HTTP Metrics exposed by Prometheus** and **List of Metrics not exposed by Prometheus** maybe. Currently it's not clear and looks like two lists of the same thing.

Also, **Histogram** is bunched in with **Timer**. Given it appears to be a separate Type it would be good to give it it's own definition line, as skimming up to check could mean missing it.

Also, in the list of **Common HTTP Metrics** what is the difference between the little ones, like http_verb, and the longer ones; it's not clear if there is a difference but the different header size/style suggests there is.

**Comment by stefanobaghino-da** [comment]
For reference, is there a (recent) PR where this functionality was added but left undocumented? Or did we have it for a long time in an undocumented status?

File: docs/source/tools/trigger-service/index.rst (hunk 1 above)
Comment (miklos-da) [reply docs/source/tools/trigger-service/index.rst:365]: Good point — changed.

**Comment by miklos-da** [comment]
> From a language perspective, it would be good to clarify the difference between the two headers **Common HTTP Metrics** and **List of Metrics** if there is some significant difference between them. Or join them if there isn't. You can do this in the title, i.e. **Common HTTP Metrics exposed by Prometheus** and **List of Metrics not exposed by Prometheus** maybe. Currently it's not clear and looks like two lists of the same thing.

I've moved the common HTTP metrics into a separate section under `ops/common-metrics` and referenced it from there from the pages for both HTTP-JSON services.

> Also, **Histogram** is bunched in with **Timer**. Given it appears to be a separate Type it would be good to give it it's own definition line, as skimming up to check could mean missing it.

I've removed mention of the histogram from there as it is indeed confusing. We should be able to completely remove this section if we simply referenced the Prometheus metric definitions.
Also, if we auto-generated metrics documentation for the json-api-service as well (just like e.g. for the Ledger API Server), we didn't need to maintain the `Metrics Reference` section.

> Also, in the list of **Common HTTP Metrics** what is the difference between the little ones, like http_verb, and the longer ones; it's not clear if there is a difference but the different header size/style suggests there is.

Changed the formatting to a bulleted list.

File: docs/source/ops/common-metrics.rst [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+- **host**: fully qualified hostname of the HTTP endpoint (e.g. ``example.com``)
+
+- **path**: path of the HTTP endpoint (e.g. ``/parties/create``)
+
+- **service**: Daml service's name (``json-api`` for the HTTP JSON API Service)
+
+daml.http.requests.duration.seconds
+===================================
+- **Description**: Records the durations of serving HTTP requests.
+- **Type**: Histogram
```
[the remark this answers is not in the document]
Comment (katmurp) [reply docs/source/ops/common-metrics.rst:27]: see: metrics.rst

**Review by katmurp — COMMENTED** [review]
Just one comment about Histogram type not being in the Type list.

File: docs/source/ops/common-metrics.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (katmurp) [reply docs/source/ops/common-metrics.rst:27]: SGTM.

File: docs/source/ops/common-metrics.rst [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+===========================================
+- **Description**: Records payload sizes of WebSocket messages received.
+- **Type**: Histogram
+
+daml.http.websocket.messages.sent.bytes
+=======================================
+- **Description**: Records payload sizes of WebSocket messages sent.
+- **Type**: Histogram
+
+
```
[the remark this answers is not in the document]
Comment (katmurp) [reply docs/source/ops/common-metrics.rst:68]: I can't answer this as I don't have a broad enough understanding of what's happening here.

**Review by lucb-da — APPROVED** [review]
The changes are fine by me.

I am not a big fan of using verbs like `records` and `counts`, it feels like it explains how the metrics are created, instead of what they represent.
I prefer something more descriptive of what the value is, like `Total number of ...`.

I had a look at the current metrics documentation for Canton, it is a mix of both styles.
We need to add it to list of details to review and standardize.

**Comment by miklos-da** [comment]
> The changes are fine by me.
> 
> I am not a big fan of using verbs like `records` and `counts`, it feels like it explains how the metrics are created, instead of what they represent. I prefer something more descriptive of what the value is, like `Total number of ...`.
> 
> I had a look at the current metrics documentation for Canton, it is a mix of both styles. We need to add it to list of details to review and standardize.

Thanks for your feedback, @lucb-da. I have replaced mentions of `records` and `counts` as suggested.

File: docs/source/ops/common-metrics.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (miklos-da) [reply docs/source/ops/common-metrics.rst:140]: I have moved the common gRPC metric definitions here.

File: docs/source/ops/common-metrics.rst [hunk 4]
```diff
@@ -5,30 +5,102 @@ Common Metrics
...
+daml.grpc.server.started.total
+==============================
+- **Description**: Total number of started gRPC requests (on either type of connection).
+- **Type**: Counter
+
+.. errors
+
+daml.grpc.server.handled.total
+==============================
+- **Description**: Counts number of handled gRPC requests.
```
Comment (lucb-da) [inline docs/source/ops/common-metrics.rst:60]: One more `counts` which could be replaced.
```suggestion
- **Description**: Total number of handled gRPC requests.
```

**Pull request merged**